open_spiel/games/battleship.cc

```diff
@@ -0,0 +1,887 @@
+      absl::StrAppend(&state_str, absl::get<Shot>(move.action).ToString());
+    }
+  }
+  return state_str;
+}
+
+std::string BattleshipState::ToPrettyString() const {
+  std::string state_str;
+
+  absl::StrAppend(&state_str, "Player 0's board:\n");
+  absl::StrAppend(&state_str, OwnBoardString(Player{0}));
+  absl::StrAppend(&state_str, "Player 1's board:\n");
+  absl::StrAppend(&state_str, OwnBoardString(Player{1}));
+  // NOTE: It is not necessary to also include the boards of the players' shots,
+  //     since those can be inferred from the two ship boards.
+
+  return state_str;
+}
+
+bool BattleshipState::IsTerminal() const {
+  return CurrentPlayer() == kTerminalPlayerId;
+}
+
+std::vector<double> BattleshipState::Returns() const {
+  SPIEL_CHECK_TRUE(IsTerminal());
+  const BattleshipConfiguration& conf = bs_game_->configuration;
+
+  // The description of the game in the header file contains more details
+  // about how the payoffs for the players are computed at the end of the
+  // game, as well as the meaning of the `loss_multiplier`.
+  const double loss_multiplier = conf.loss_multiplier;
+
+  double damage_pl1 = 0.0;
+  double damage_pl2 = 0.0;
+  for (const Ship& ship : conf.ships) {
+    if (DidShipSink(ship, Player{0})) damage_pl1 += ship.value;
+    if (DidShipSink(ship, Player{1})) damage_pl2 += ship.value;
+  }
+
+  return {damage_pl2 - loss_multiplier * damage_pl1,
+          damage_pl1 - loss_multiplier * damage_pl2};
+}
+
+std::unique_ptr<State> BattleshipState::Clone() const {
+  return std::make_unique<BattleshipState>(*this);
+}
+
+std::string BattleshipState::InformationStateString(Player player) const {
+  SPIEL_CHECK_TRUE(player >= 0 && player < NumPlayers());
+
+  const BattleshipConfiguration& conf = bs_game_->configuration;
+  const Player opponent = (player == Player{0}) ? Player{1} : Player{0};
+
+  // We will need to figure out whether each of the player's shots (i) hit the
+  // water, (ii) damaged but did not sink yet one of the opponent's ships, or
+  // (iii) damaged and sank one of the opponent's ships.
+  //
+  // To be able to figure that out, we will keep track of the damage that each
+  // of the opponent's ship has received so far. The vector `ship_damage`
+  // contains and updates this information as each player's shot is processed
+  // in order. Position i corresponds to the damage that the opponent's ship
+  // in position i of bs_game->configuration.ships has suffered.
+  std::vector<int> ship_damage(conf.ships.size(), 0);
+  // Since in general we might have repeated shots, we cannot simply increase
+  // the ship damage every time a shot hits a ship. For that, we keep track of
+  // whether a cell was already hit in the past. We reuse the
+  // serialization/deserialization routines for shots to map from (r, c) to
+  // cell index r * board_width + c.
+  std::vector<bool> cell_hit(conf.board_width * conf.board_height, false);
+
+  std::string information_state;
```

Comment:
Interesting, what is the rationale there?

Reply:
FOG games are 1-timeable, meaning the "flow of time" is a publicly observable event. Think of it as each agent sees a clock on the wall. The consistency error you were getting said that the information state was spanning multiple time steps (was not subset of MoveNumber). The suggested code is an easy way to fix it, though I think it is generally better to do it via encoding the time implicitly from the game logic. Then the test would reveal if you have indeed grasped and encoded the logic correctly (consistently).

Reply:
That makes sense but I'm not sure I fully see it yet why the previous string didn't already encode the MoveNumber implicitly? Isn't that just a function of how many actions have happened so far?

Reply:
Not sure why in this game. In some games it happens that the observations were empty until the setup phase is finished, and the opponent players didn't receive appropriate observations. So maybe in this game the observations for the opponent should have been something like "player one is setting up first ship", "player one is setting up second ship", "player two is setting up his first ship, and I, player 1, have ships at positions X,Y", "player two is setting up his second ship, and I, player 1, have ships at positions X,Y" etc. 

By inserting MoveNumber() it results in the same partition of the history tree, though the string description is different (and less human-understandable)

Reply:
Ah, I think I get it now. For example: Player 2 cannot distinguish between "Player 1 still hasn't placed the first ship" and "Player 1 has placed the ship and it's now my turn" in the first two moves.

Reply:
Yes, I agree we should update the docs at some point. The observer API is recent, but I think there shouldn't be major changes to it anymore. Also I didn't make the consistency test for tensor observations yet, that will be quite a "juicy" PR as well.  The issue is that the original API required to implement observations only when the player is acting, but for the purposes of public state decomposition we need to have them for any position in the game.

Reply:
Just a clarification: the original API does allow observations at any point, this is why the functions have a player parameter. Many of the games put this observing player id in the observation to distinguish whose turn it is versus who is observing the state.

Reply:
Perfect. I've added a note in the code, and will keep it in mind for the next game :-)  I'm going to resolve this for now. Since this requirement about timeability is not immediately intuitive (and in fact several games do not respect it), perhaps as a first step it would be worth suggesting to check whether the game is 1-timeable when the consistency checks fail.

Thanks for the help. I'm going to resolve this for now.